crux_core/src/steps.rs

```diff
@@ -1,143 +1,107 @@
-use std::{
-    collections::{hash_map::Entry, HashMap},
-    fmt,
-    sync::Mutex,
-};
+use std::fmt::{self, Debug};
+use thiserror::Error;
 
-use uuid::Uuid;
+use crate::capability::Operation;
 
-use crate::Request;
-
-pub(crate) struct Step<T> {
-    pub(crate) payload: T, // T is an Operation first, then Effect once mapped
-    pub(crate) resolve: Option<Resolve>,
-}
-
-type OnceFn = dyn FnOnce(&[u8]) + Send;
-type ManyFn = dyn Fn(&[u8]) -> Result<(), ()> + Send;
-
-pub(crate) enum Resolve {
-    Once(Box<OnceFn>),
-    Many(Box<ManyFn>),
-}
+pub struct Step<Op>(pub Op, pub Resolve<Op::Output>)
+where
+    Op: Operation;
 
-impl<T> Step<T> {
-    pub fn resolves_never(payload: T) -> Self {
-        Self {
-            payload,
-            resolve: None,
-        }
+impl<Op> Step<Op>
+where
+    Op: Operation,
+{
+    pub(crate) fn resolves_never(payload: Op) -> Self {
+        Self(payload, Resolve::Never)
     }
 
-    pub fn resolves_once<F>(payload: T, resume: F) -> Self
+    pub(crate) fn resolves_once<F>(payload: Op, resolve: F) -> Self
     where
-        F: FnOnce(&[u8]) + Send + 'static,
+        F: FnOnce(Op::Output) + Send + 'static,
     {
-        Self {
-            payload,
-            resolve: Some(Resolve::Once(Box::new(resume))),
-        }
+        Self(payload, Resolve::Once(Box::new(resolve)))
     }
 
-    pub fn resolves_many_times<F>(payload: T, resume: F) -> Self
+    pub(crate) fn resolves_many_times<F>(payload: Op, resolve: F) -> Self
     where
-        F: Fn(&[u8]) -> Result<(), ()> + Send + 'static,
+        F: Fn(Op::Output) -> Result<(), ()> + Send + 'static,
     {
-        Self {
-            payload,
-            resolve: Some(Resolve::Many(Box::new(resume))),
-        }
+        Self(payload, Resolve::Many(Box::new(resolve)))
     }
 
-    pub fn map_effect<Ef, F>(self, f: F) -> Step<Ef>
-    where
-        F: Fn(T) -> Ef + Sync + Send + Copy + 'static,
-        T: 'static,
-        Ef: 'static,
-    {
-        Step {
-            payload: f(self.payload),
-            resolve: self.resolve,
-        }
+    pub(crate) fn resolve(&mut self, output: Op::Output) -> Result<(), ResolveError> {
+        self.1.resolve(output)
     }
 }
 
-impl<T> fmt::Debug for Step<T>
+impl<Op> fmt::Debug for Step<Op>
 where
-    T: fmt::Debug,
+    Op: Operation + Debug,
 {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Step")
-            .field("payload", &self.payload)
-            .finish_non_exhaustive()
+        f.debug_tuple("Step").field(&self.0).finish()
     }
 }
 
-type Store = HashMap<[u8; 16], Resolve>;
-
-pub(crate) struct StepRegistry(Mutex<Store>);
-
-impl Default for StepRegistry {
-    fn default() -> Self {
-        Self(Mutex::new(Store::new()))
+impl<Op> PartialEq for Step<Op>
+where
+    Op: Operation,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
```

Comment:
Guessing this is why you've made `Operation: PartialEq` elsewhere, but still wondering why `Step<Op>` needs PartialEq?